### DISCUSSION TITLE
Enabling DPD traffic even when there is a traffic from peer

### DISCUSSION THREAD

**Discussion by piddalagi** [strongswan/strongswan#2271] [Q&A]
Dear Team,

I have some use-case where I want to send keep-alive traffic periodically even when there is traffic being received from peer. Is it possible in strongswan to achieve that?

**Comment by tobiasbrunner** [comment]
> keep-alive traffic

What kind of keep-alive traffic are you referring to exactly? And what's the use case?

> Is it possible in strongswan to achieve that?

No, NAT keepalives and DPDs are both not sent if there is traffic (outbound or inbound, respectively). You might want to consider sending regular traffic through the actual tunnel(s) (e.g. ping some remote host).

**Reply by piddalagi** [reply]
My use-case is to leverage DPD traffic to monitor the tunnel health irrespective of presence of outbound/inbound traffic and not depend on external traffic like ping. If there is no existing knob for the same, do you suggest me to tweak the code accordingly?

**Reply by tobiasbrunner** [reply]
> My use-case is to leverage DPD traffic to monitor the tunnel health irrespective of presence of outbound/inbound traffic

What do you mean exactly?

**Reply by piddalagi** [reply]
I am using IKEv2.
When there is no inbound/outbound traffic, the the tunnels are checked for liveliness using below message exchanges.

18:45:01.147489 IP 10.0.1.1.4500 > 20.0.1.2.4500: NONESP-encap: isakmp: child_sa  inf2
18:45:01.149583 IP 20.0.1.2.4500 > 10.0.1.1.4500: NONESP-encap: isakmp: child_sa  inf2[I]
18:45:01.150869 IP 10.0.1.1.4500 > 20.0.1.2.4500: NONESP-encap: isakmp: child_sa  inf2[R]
18:45:01.150916 IP 20.0.1.2.4500 > 10.0.1.1.4500: NONESP-encap: isakmp: child_sa  inf2[IR]

We see those messages every 30s as the DPD delay is configured with that value.

Now my requirement is that I want to schedule liveliness-check irrespective of presence/absence of inbound/outbound traffic.

**Reply by tobiasbrunner** [reply]
> Now my requirement is that I want to schedule liveliness-check irrespective of presence/absence of inbound/outbound traffic.

Yes, but to what end?

**Reply by piddalagi** [reply]
Sorry shall I know what do you mean by to what end?
I want have this check periodically, say every x seconds.

**Reply by tobiasbrunner** [reply]
Yeah, but why? What's the purpose?

**Reply by piddalagi** [reply]
I want to use these packets in measuring the one-way-delay, which will be used as one of the factors in assessing the SLA.

**Reply by tobiasbrunner** [reply]
Hm, DPDs are small UDP packets sent between the IKE daemons. That might not accurately represent the delay for actual IPsec traffic (in particular if there is no UDP encapsulation, but even then the processing paths of the packets on the end systems will be pretty different). For that you really have to send traffic through the tunnel.

**Reply by piddalagi** [reply]
Thanks for the insight. 
Shall I know why the packet INFORMATIONAL goes unencrypted?

**Reply by tobiasbrunner** [reply]
What do you mean?

**Reply by piddalagi** [reply]
I have stopped all traffic from both left and right devices on which IPSec is configured, because I wanted to see the keepalive traffic after DPD delay.
I was thinking that these packets will also be tunneled. So wanted to know why they are not tunneled? 

18:45:01.147489 IP 10.0.1.1.4500 > 20.0.1.2.4500: NONESP-encap: isakmp: child_sa inf2
18:45:01.149583 IP 20.0.1.2.4500 > 10.0.1.1.4500: NONESP-encap: isakmp: child_sa inf2[I]
18:45:01.150869 IP 10.0.1.1.4500 > 20.0.1.2.4500: NONESP-encap: isakmp: child_sa inf2[R]
18:45:01.150916 IP 20.0.1.2.4500 > 10.0.1.1.4500: NONESP-encap: isakmp: child_sa inf2[IR]

**Reply by tobiasbrunner** [reply]
As I already wrote [above](https://github.com/strongswan/strongswan/discussions/2271#discussioncomment-9660002), the DPDs are IKE packets exchanged between the IKE daemons. Encrypted and authenticated, but not sent through an IPsec SA, as there might not even be one.

**Reply by piddalagi** [reply]
Tobi, thank you for the information. We can close this discussion thread.
